joomla/joomla-cms
# [4.0] Change collations of com_finder tables from general to unicode on MySQL databases.

Pull Request for Issue #9361 .

Redo of Pull Request (PR) #16617 but for 4.0 and in a different way.

### Summary of Changes

This Pull Request (PR) changes collations of database tables for com_finder from general to unicode collations for MySQL databases.

We had left them with general collations when we did the utf8mb4 conversion because otherwise things broke when tables from 3rd party extensions where joined to com_finder tables.

Now with J4 com_finder database tables have been restructured anyway, so most of the tables are either dropped and created again in the update sql script changed with this PR, or they are at least truncated, so it is a good chance to change the table collations now without having to do what was before tried with PR #16617 for that purpose, which was a kind of 2nd stage of the utf8mb4 conversion procedure. Especially the `#__finder_terms` table required a special conversion in PR #16617 because of possible duplicate records after conversion, but now in J4 this is not a problem anymore, because the `#__finder_terms` table is truncated anyway in the update sql, also before this PR here. That's why it now can be done just with the update sql script.

For PostgreSQL databases, it adds truncation of tables `#__finder_taxonomy_map`, `#__finder_tokens` and `#__finder_tokens_aggregate` in the same way as it adds it for MySQL databases.

### Testing Instructions

All tests related to collations have to be done using a MySQL (or MariaDB) database.

For PostgreSQL nothing has to be done for new installation test. For the update test, just check as described if everything works as well as before.

There are 2 tests to be done:
1. New installation
2. Update

#### Instructions for new installation test

1. On a current 4.0dev branch, apply the patch of this PR and make a new installation. If you are not woorking on a git clone you can use following installation package, which is equal to last nightly build plus the patch of this PR applied: [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Full_Package_2020-03-23_pr-28425.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Full_Package_2020-03-23_pr-28425.zip).
2. When the update has finished, log in to backend and check that everything works.
3. Check in your database using e.g. PhpMyAdmin that all database tables have collation `utf8mb4_unicode_ci` except of table `#__finder_terms_common`, which has collation `utf8mb4_bin`.

#### Instructions for update test

1. Install a fresh copy of 3.10 nightly or a git clone of the current 3.10-dev branch and install some sample data, or have a 3.9 installation with much content and update that one to 3.10.
3. Enable the Smart Search plugin, if not enabled yet, enable statistics in Smart Search options and use smart search in frontend. Make sure you have indexed content and statistics.
4. Go to the Joomla Update Component component options, set the update channel to "Custom ULR", set the URL to [https://test5.richard-fath.de/next_major_list_pr-28425.xml](https://test5.richard-fath.de/next_major_list_pr-28425.xml) and the status to "Development".
5. Update to Joomla 4.0.
6. When the update has finished, log in to backend and check that everything works. There is an error message shown about not existing template style, this is known and happens also without this PR. Beside this there should not be any error, and no errors or warnings or notices should be shown in PHP error log.
7. Check in your database using e.g. PhpMyAdmin that all database tables have collation `utf8mb4_unicode_ci` except of table `#__finder_terms_common`, which has collation `utf8mb4_bin`.

### Expected result

All database tables of the Joomla CMS core have collation `utf8mb4_unicode_ci` except of table `#__finder_terms_common`, which has collation `utf8mb4_bin`.

### Actual result

All database tables of the Joomla CMS core have collation `utf8mb4_unicode_ci` except of tables with names starting with `#__finder`. Those have either collation `utf8mb4_general_ci` or `utf8mb4_bin`.

### Documentation Changes Required

None.
